node/src/transaction_pool/transaction_pool_reducer.rs

```diff
@@ -315,22 +307,37 @@ impl TransactionPoolState {
                 };
 
                 let dispatcher = state.into_dispatcher();
-                if let Some(rpc_action) = rpc_action {
-                    dispatcher.push(rpc_action);
+
+                // TODO: use callbacks
+                match (was_accepted, from_source) {
+                    (true, TransactionPoolMessageSource::Rpc { id }) => {
+                        dispatcher.push(RpcAction::TransactionInjectSuccess {
+                            rpc_id: id,
+                            response: accepted.clone(),
+                        });
+                    }
+                    (true, TransactionPoolMessageSource::Pubsub { id }) => {
+                        dispatcher.push(P2pNetworkPubsubAction::BroadcastValidatedMessage {
+                            message_id: BroadcastMessageId::MessageId { message_id: id },
+                        });
+                    }
+                    (false, TransactionPoolMessageSource::Rpc { id }) => {
+                        dispatcher.push(RpcAction::TransactionInjectRejected {
+                            rpc_id: id,
+                            response: rejected.clone(),
+                        });
+                    }
+                    (false, TransactionPoolMessageSource::Pubsub { id }) => {
+                        dispatcher.push(P2pNetworkPubsubAction::RejectMessage {
+                            message_id: Some(BroadcastMessageId::MessageId { message_id: id }),
+                            peer_id: None,
+                            reason: "Rejected transaction diff".to_owned(),
+                        });
+                    }
+                    (_, TransactionPoolMessageSource::None) => {}
                 }
 
                 if was_accepted {
```

Comment:
should this be a match above for `(true, TransactionPoolMessageSource::None)` instead? we don't want rebroadcast if we pushed any of the `P2pNetworkPubsubAction` actions above right?

Reply:
Actually, it should happen for the RPC source too. Just not for the case in which the source is the libp2p gossip